### PR TITLE
Add matchesPattern support

### LINKED ISSUE #2304: An unknown function with name 'matchesPattern' was found.
I'm using Microsoft.AspNet.OData.

I'm trying to use $filter=matchesPattern(Property, '%pattern%') but I'm receiving Exception: Microsoft.OData.ODataException: 'An unknown function with name 'matchesPattern' was found. This may also be a function import or a key lookup on a navigation property, which is not allowed.'

Is 'matchesPattern' function not supported?

It is a [specification](http://docs.oasis-open.org/odata/odata/v4.01/odata-v4.01-part2-url-conventions.html#sec_matchesPattern) of OData v4 protocol

### PR DESCRIPTION
**Pull request by au5ton**
Based on code by @Arithmomaniac in https://github.com/OData/AspNetCoreOData/pull/729

### Issues

This pull request fixes #2304. 

### Description

In short, this pull request adds support for the `matchesPattern` string URI function.

This pull request aims to bring the changes made in https://github.com/OData/AspNetCoreOData/pull/729 to this version of the project. The code is based heavily on that pull request created by @Arithmomaniac.

### Checklist (Uncheck if it is not completed)

- [X] *Test cases added*
- [X] *Build and test with one-click build and test script passed*

### CODE REVIEW

**Comment by au5ton** [comment]
@microsoft-github-policy-service agree

File: src/Microsoft.AspNet.OData.Shared/Query/Expressions/ClrCanonicalFunctions.cs [hunk 1]
```diff
@@ -61,6 +63,7 @@ internal class ClrCanonicalFunctions
         public static readonly MethodInfo ToUpper = MethodOf(_ => _defaultString.ToUpper());
         public static readonly MethodInfo Trim = MethodOf(_ => _defaultString.Trim());
         public static readonly MethodInfo Concat = MethodOf(_ => String.Concat(default(string), default(string)));
+        public static readonly MethodInfo MatchesMattern = MethodOf(_ => Regex.IsMatch(default(string), default(string), default(RegexOptions)));
```
Comment (Arithmomaniac) [inline src/Microsoft.AspNet.OData.Shared/Query/Expressions/ClrCanonicalFunctions.cs:66]: Name typo - `MatchesPattern`

Comment (au5ton) [reply src/Microsoft.AspNet.OData.Shared/Query/Expressions/ClrCanonicalFunctions.cs]: Thank you, for catching this. I addressed this in f5f20a8cdd4ab96fa280136bb7ad3ca6b00ec67f.

**Review by KenitoInc — APPROVED** [review]

**Review by gathogojr — APPROVED** [review]
:shipit:

**Pull request merged**